src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1796,9 +1784,19 @@ private void initCEN(int knownTotal) throws IOException {
                         // performance in multi-release jar files
                         int version = getMetaVersion(entryPos + META_INF_LEN, nlen - META_INF_LEN);
                         if (version > 0) {
-                            if (metaVersionsSet == null)
-                                metaVersionsSet = new TreeSet<>();
-                            metaVersionsSet.add(version);
+                            try {
+                                // Compute hash code of name from "META-INF/versions/{version)/{name}
+                                int prefixLen = META_INF_VERSIONS_LEN + DecimalDigits.stringSize(version);
+                                int hashCode = zipCoderForPos(pos).checkedHash(cen,
+                                        entryPos + prefixLen,
+                                        nlen - prefixLen);
+                                // Register version for this hash code
+                                if (metaVersions == null)
+                                    metaVersions = new HashMap<>();
+                                metaVersions.computeIfAbsent(hashCode, _ -> new BitSet()).set(version);
+                            } catch (Exception e) {
+                                throw new IllegalArgumentException(e);
```

Comment:
Hello Eirik, I'm late to this to PR. Was throwing an unspecified `IllegalArgumentException` from here intentional? Should it have been `IOException` instead?

Reply:
@eirbjo will know for sure but I think it's intentional since if there's an encoding error we'd have thrown earlier when hashing over the full name.

Reply:
Hello Claes, 

> if there's an encoding error we'd have thrown earlier when hashing over the full name.

If I understand correctly, then I think you are referring to the `ZipCoder.checkedHash()` which throws that `Exception`.

Before the change in this PR, there was (and continues to exist) one such call in the `checkAndAddEntry()` private method of this `Source` class. In that place we catch the `Exception` and throw a `ZipException` (which is an `IOException`) https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/zip/ZipFile.java#L1264.

I think we should be doing the same here and throw a `ZipException` instead of an `IllegalArgumentException`

Reply:
What I mean is this IAE cant be thrown and is more of an assert. But sure, it should throw a ZipException to be more robust and obviously correct.

Reply:
Yes, this was a "this will never happen" situation so I thought of it as an assert. But that's not locally obvious and I agree it's better to be consistent and use ZipException here.

See #21791 for a fix.